system-parachains/coretime/coretime-kusama/src/coretime.rs

```diff
@@ -0,0 +1,280 @@
+use frame_support::{
+	parameter_types,
+	traits::{
+		fungible::{Balanced, Credit},
+		OnUnbalanced,
+	},
+};
+use pallet_broker::{
+	AdaptPrice, CoreAssignment, CoreIndex, CoretimeInterface, PartsOf57600, RCBlockNumberOf,
+};
+use parachains_common::{AccountId, Balance, BlockNumber};
+use sp_runtime::{
+	traits::{One, Saturating},
+	FixedU64,
+};
+use xcm::latest::prelude::*;
+
+pub struct CreditToCollatorPot;
+impl OnUnbalanced<Credit<AccountId, Balances>> for CreditToCollatorPot {
+	fn on_nonzero_unbalanced(credit: Credit<AccountId, Balances>) {
+		let staking_pot = CollatorSelection::account_id();
+		let _ = <Balances as Balanced<_>>::resolve(&staking_pot, credit);
+	}
+}
+
+/// A type containing the encoding of the coretime pallet in the Relay chain runtime. Used to
+/// construct any remote calls. The codec index must correspond to the index of `Coretime` in the
+/// `construct_runtime` of the Relay chain.
+#[derive(Encode, Decode)]
+enum RelayRuntimePallets {
+	#[codec(index = 74)]
+	Coretime(CoretimeProviderCalls),
+}
+
+/// Call encoding for the calls needed from the relay coretime pallet.
+#[derive(Encode, Decode)]
+enum CoretimeProviderCalls {
+	#[codec(index = 1)]
+	RequestCoreCount(CoreIndex),
+	#[codec(index = 2)]
+	RequestRevenueInfoAt(relay_chain::BlockNumber),
+	#[codec(index = 3)]
+	CreditAccount(AccountId, Balance),
+	#[codec(index = 4)]
+	AssignCore(
+		CoreIndex,
+		relay_chain::BlockNumber,
+		Vec<(CoreAssignment, PartsOf57600)>,
+		Option<relay_chain::BlockNumber>,
+	),
+}
+
+parameter_types! {
+	pub const BrokerPalletId: PalletId = PalletId(*b"py/broke");
+}
+
+parameter_types! {
+	pub storage CoreCount: Option<CoreIndex> = None;
+	pub storage CoretimeRevenue: Option<(BlockNumber, Balance)> = None;
+}
+
+/// Type that implements the `CoretimeInterface` for the allocation of Coretime. Meant to operate
+/// from the parachain context. That is, the parachain provides a market (broker) for the sale of
+/// coretime, but assumes a `CoretimeProvider` (i.e. a Relay Chain) to actually provide cores.
+pub struct CoretimeAllocator;
+impl CoretimeInterface for CoretimeAllocator {
+	type AccountId = AccountId;
+	type Balance = Balance;
+	type RealyChainBlockNumberProvider = RelaychainDataProvider<Runtime>;
+
+	fn request_core_count(count: CoreIndex) {
+		use crate::coretime::CoretimeProviderCalls::RequestCoreCount;
+		let request_core_count_call = RelayRuntimePallets::Coretime(RequestCoreCount(count));
+
+		// Weight for `request_core_count` from Kusama runtime benchmarks:
+		// `ref_time` = 7889000 + (3 * 25000000) + (1 * 100000000) = 182889000
+		// `proof_size` = 1636
+		// Add 5% to each component and round to 2 significant figures.
+		let call_weight = Weight::from_parts(190_000_000, 1700);
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
```

Comment:
I'm surprised this is `Native`, I would have expected `Xcm` with a mapping to some origin or `BodyId`. So just sanity checking, this is what we use on Rococo?

Reply:
Yes this is what we used on Rococo, these calls need to have the origin of the coretime chain (which is what `Native` converts to in this case)